tpkg/cutest/cutest_event.c

```diff
@@ -0,0 +1,364 @@
+	(void)discard;
+	exit(0);
+}
+
+static void
+agent_callback(int fd, short event, void *arg)
+{
+	char buf[] = "x";
+	struct proc *proc = (struct proc *)arg;
+	ssize_t discard;
+
+	(void)fd;
+	assert(arg != NULL);
+
+	if (event & EV_TIMEOUT) {
+		discard = write(fd, buf, strlen(buf));
+		(void)discard;
+	} else if(event & EV_WRITE) {
+		discard = write(fd, buf, strlen(buf));
+		(void)discard;
+		proc->events |= EV_WRITE;
+	} else {
+		assert(event & EV_CLOSED);
+		proc->events |= EV_CLOSED;
+	}
+}
+
+static void
+child_callback(int fd, short event, void *arg)
+{
+	struct proc *proc = (struct proc *)arg;
+
+	(void)fd;
+	assert(arg != NULL);
+
+	if (event & EV_READ) {
+		proc->events |= EV_READ;
+	} else {
+		assert(event & EV_CLOSED);
+		proc->events |= EV_CLOSED;
+	}
+}
+
+static void
+sigchld_callback(int sig, short event, void *arg)
+{
+	pid_t pid;
+	int wstatus = 0;
+	struct testvars *vars = (struct testvars *)arg;
+	struct proc *proc;
+
+	assert(sig == SIGCHLD);
+	assert(event & EV_SIGNAL);
+	assert(arg != NULL);
+
+	do {
+		pid = waitpid(-1, &wstatus, WNOHANG);
+		proc = NULL;
+		if(vars->agent.pid == pid) {
+			proc = &vars->agent;
+		} else if(pid > 0) {
+			for(int i = 0; i < NUM_CHILDREN; i++) {
+				if(vars->children[i].pid == pid) {
+					proc = &vars->children[i];
+				}
+			}
+		}
+		if(proc != NULL) {
+			vars->sigchld_count++;
+			proc->wstatus = wstatus;
+			proc->events |= EV_SIGNAL;
+		}
+	} while((pid == -1 && errno == EINTR) || (pid > 0));
+
+	if(vars->sigchld_count == (NUM_CHILDREN + 1)) {
+		event_base_loopexit(vars->event_base, NULL);
+	}
+}
+
+static void
+fork_children(struct testvars *vars)
+{
+	for(int i = 0; i < NUM_CHILDREN; i++) {
+		pid_t pid;
+		int ret, fdin[2], fdout[2];
+		ret = pipe(fdin);
+		assert(ret == 0);
+		ret = pipe(fdout);
+		assert(ret == 0);
+		if((pid = fork()) == 0) {
+			close(fdin[1]);
+			close(fdout[0]);
+			child(fdin[0], fdout[1]);
+			/* never reached */
+		}
+		close(fdin[0]);
+		close(fdout[1]);
+		vars->children[i].pid = pid;
+		vars->children[i].fdin = fdin[1];
+		vars->children[i].fdout = fdout[0];
+		event_set(
+			&vars->children[i].event,
+			 vars->children[i].fdin,
+			 EV_READ | EV_CLOSED,
+			&child_callback,
+			&vars->children[i]);
+		event_base_set(vars->event_base, &vars->children[i].event);
+		event_add(&vars->children[i].event, NULL);
+	}
+}
+
+static void
+fork_agent(struct testvars *vars, void(*func)(struct proc *))
+{
+	pid_t pid;
+	int ret, fdin[2], fdout[2];
+	struct timeval timeout = { 1, 0 };
+	ret = pipe(fdin);
+	assert(ret == 0);
+	ret = pipe(fdout);
+	assert(ret == 0);
+	if((pid = fork()) == 0) {
+		close(fdin[1]);
+		close(fdout[0]);
+		agent(fdin[0], fdout[1], vars, func);
+		/* never reached */
+	}
+	close(fdin[0]);
+	close(fdout[1]);
+	vars->agent.pid = pid;
+	vars->agent.fdin = fdin[1];
+	vars->agent.fdout = fdout[0];
+	event_set(
+		&vars->agent.event,
+		 vars->agent.fdin,
+		 EV_WRITE | EV_CLOSED,
```

Comment:
Does EV_CLOSED really work for other event backends, like select(2) (or with mini_event) ?  Because I guess it may be epoll or kqueue specific.

Reply:
Indeed. Though for select it could be supported, but I've removed it for now.